### DISCUSSION TITLE
I need help with surfaceFromPoint and update Hotspot.

### DISCUSSION THREAD

**Discussion by bigbea1r** [google/model-viewer#4845] [Q&A] [answered]
I need this code so that I can drag the Hotspot with the mouse over the entire model, and at the same time so that the hotspot is on the surface of the model. And for this, you need the data-surface attribute. I use it when creating a hotspot. but to drag, I need to update. At the moment, the code is able to update the hotspot, but as soon as I start dragging it, the hotspot disappears immediately, but the html code remains unchanged and the element itself remains in html. Please help me figure out what the problem is.
Here is a part of the code:
```
function startDrag(event) {
  if (editMode) {
    const hotspot = event.target;

    async function onMouseMove(moveEvent) {
      const newSurface = modelViewer.surfaceFromPoint(moveEvent.clientX, moveEvent.clientY);
      if (newSurface) {
        hotspot.setAttribute('data-surface', newSurface);
        console.log(`Updated data-surface: ${newSurface}`);
        await modelViewer.updateComplete;

        const updatedSurface = {
           name: currentSlot,
           position: `${newSur}`
         };
         modelViewer.updateHotspot(updatedSurface);
      }
    }

    function onMouseUp() {
      document.removeEventListener('mousemove', onMouseMove);
      document.removeEventListener('mouseup', onMouseUp);
      saveButtons(); // Сохранение изменений
    }

    document.addEventListener('mousemove', onMouseMove);
    document.addEventListener('mouseup', onMouseUp);
  }
}
```
Here is the code for creating and assigning a hotspot location:
```
function createHotspot(animationName, surface, text) {
  const hotspot = document.createElement('button');
  const hotspotDiv = document.createElement('div');

  animNameCreate.push(animationName);   
  hotspot.setAttribute('class', 'Hotspot');   
  hotspot.setAttribute('slot', `hotspot-${animNameCreate.length}`);   
  hotspot.setAttribute('data-surface', surface);   
  hotspot.setAttribute('data-visibility-attribute', 'visible');   
  hotspot.setAttribute('data-animation', animationName);   
  hotspotDiv.setAttribute('class', 'HotspotAnnotation');   
  hotspotDiv.textContent = text; 
  hotspot.appendChild(hotspotDiv);

  hotspot.addEventListener('click', async () => {
    if (!editMode) {
      playAnimation(animationName, doorCondition ? -1 : 1);
      doorCondition = !doorCondition;
    } else {
      editDiv(hotspot, hotspotDiv);
    }
  });

  hotspot.addEventListener('mousedown', startDrag);

  modelViewer.appendChild(hotspot);

  saveButtons();
}
```
I left the whole code just in case:
```
import Component from './conponent.js';

let editMode = false;
let divNameCreate = []
let animNameCreate = [];
let component = new Component();
const modelViewer = document.querySelector('model-viewer');
const editButton = document.getElementById('edit_button');
const popupOn = document.getElementById('popupOn');
const popupOff = document.getElementById('popupOff');
const popupError = document.getElementById('popupError');
const containerInput = document.getElementById('containerInput');
const divNameInput = document.getElementById('divNameInput');
const animationNameInput = document.getElementById('animationNameInput');
const enterButton = document.getElementById('enter');
const cancelButton = document.getElementById('cancel');

let doorCondition = false;
let hotspotOn = false;
let createButtonOn = false;

async function playAnimation(anim, timeScale) {
  modelViewer.animationName = anim;
  modelViewer.timeScale = timeScale;
  await modelViewer.updateComplete;

  modelViewer.play({
    repetitions: 1,
    pingpong: false,
  });
}

component.visiblePopup(editButton, editMode, popupOn, popupOff)

editButton.addEventListener('click', () => {
  editMode = !editMode;
  if (editMode) {
    popupOn.style.display = 'block';
    setTimeout(() => {
      popupOn.style.display = 'none';
    }, 1000);
  } else {
    popupOff.style.display = 'block';
    setTimeout(() => {
      popupOff.style.display = 'none';
    }, 3000);
  }
});
cancelButton.addEventListener('click', outPopup);

function outPopup() {
  divNameInput.value = '';
  animationNameInput.value = '';
  containerInput.style.display = 'none';
  createButtonOn = true;
}
document.addEventListener('DOMContentLoaded', () =>{
  const buttons = JSON.parse(localStorage.getItem('buttons')) || [];
  console.log(buttons);
  buttons.forEach((button) => {
    createHotspot(button.animation, button.surface, button.text);
  });
})

//======================================================================
modelViewer.ondblclick = (event) => {
  if (editMode) {
      const surface = modelViewer.surfaceFromPoint(event.clientX, event.clientY);

      if (surface !== null && !hotspotOn) {
          containerInput.style.display = 'block';
          enterButton.removeEventListener('click', handleEnterClick);

          function handleEnterClick() {
              const divName = divNameInput.value;
              const anim = animationNameInput.value;
              const animName = modelViewer.availableAnimations;

              if (anim.trim() !== '' && animName.includes(anim)) {
                  createHotspot(anim, surface, divName);
                  saveButtons();
                  outPopup();
              } else {
                  outPopup();
                  popupError.style.display = 'block';
                  setTimeout(() => {
                      popupError.style.display = 'none';
                  }, 1500);
              }

              let allHotspot = document.querySelectorAll('.Hotspot');

              allHotspot.forEach(hotspot => {
                  hotspot.addEventListener('mousedown', startDrag);
              });
          }

          enterButton.addEventListener('click', handleEnterClick, { once: true });
      }
  }
};

//======================================================================
function createHotspot(animationName, surface, text) {
  const hotspot = document.createElement('button');
  const hotspotDiv = document.createElement('div');

  animNameCreate.push(animationName);   
  hotspot.setAttribute('class', 'Hotspot');   
  hotspot.setAttribute('slot', `hotspot-${animNameCreate.length}`);   
  hotspot.setAttribute('data-surface', surface);   
  hotspot.setAttribute('data-visibility-attribute', 'visible');   
  hotspot.setAttribute('data-animation', animationName);   
  hotspotDiv.setAttribute('class', 'HotspotAnnotation');   
  hotspotDiv.textContent = text; 
  hotspot.appendChild(hotspotDiv);

  hotspot.addEventListener('click', async () => {
    if (!editMode) {
      playAnimation(animationName, doorCondition ? -1 : 1);
      doorCondition = !doorCondition;
    } else {
      editDiv(hotspot, hotspotDiv);
    }
  });

  hotspot.addEventListener('mousedown', startDrag);

  modelViewer.appendChild(hotspot);

  saveButtons();
}
//======================================================================

//======================================================================
function editDiv(hotspot, hotspotDiv) {
if (editMode) {
  containerInput.style.display = 'block';
  divNameInput.value = hotspotDiv.textContent;
  animationNameInput.style.display = 'none';

  enterButton.addEventListener('click', function hsEnter() {
    if (editMode) {
      const divName = divNameInput.value;
      hotspotDiv.textContent = divName;
      hotspot.appendChild(hotspotDiv);
      animationNameInput.style.display = 'block';
      outPopup();

      const buttons = JSON.parse(localStorage.getItem('buttons')) || [];
      const buttonIndex = Array.from(modelViewer.querySelectorAll('.Hotspot')).indexOf(hotspot);
      if (buttonIndex !== -1) {
        buttons[buttonIndex].text = divName;
        localStorage.setItem('buttons', JSON.stringify(buttons));
      }
    }
  }, { once: true });
}
}
//======================================================================
function saveButtons() {
  const buttons = [];
  document.querySelectorAll('.Hotspot').forEach(button => {
    buttons.push({
      animation: button.getAttribute('data-animation'),
      surface: button.getAttribute('data-surface'),
      text: button.querySelector('.HotspotAnnotation').textContent,
    });
  });
  localStorage.setItem('buttons', JSON.stringify(buttons));
}
//======================================================================

//======================================================================

function startDrag(event) {
  if (editMode) {
    const hotspot = event.target;

    async function onMouseMove(moveEvent) {
      const newSurface = modelViewer.surfaceFromPoint(moveEvent.clientX, moveEvent.clientY);
      if (newSurface) {
        hotspot.setAttribute('data-surface', newSurface);
        console.log(`Updated data-surface: ${newSurface}`);
        await modelViewer.updateComplete;

        const updatedSurface = {
           name: currentSlot,
           position: `${newSur}`
         };
         modelViewer.updateHotspot(updatedSurface);
      }
    }

    function onMouseUp() {
      document.removeEventListener('mousemove', onMouseMove);
      document.removeEventListener('mouseup', onMouseUp);
      saveButtons(); // Сохранение изменений
    }

    document.addEventListener('mousemove', onMouseMove);
    document.addEventListener('mouseup', onMouseUp);
  }
}
//======================================================================

containerInput.addEventListener('click', (event) => {
  event.stopPropagation();
});
```
thank you in advance

**Comment by elalish** [comment]
Sorry, this looks like HTML/JS debugging, not model-viewer. If you can find a minimal repro of a model-viewer bug, I'll be happy to take a look, but I can't just debug your own website.

**Reply by bigbea1r** [reply]
The problem is that I can't find the error. In fact, it does not exist.  The problem is that hotspot does not move dynamically using data-surface. the video uses .surfaceFromPoint and .positionAndNormalFromPoint (for clarity). I need to be able to drag a hotspot, as shown in the video. The only thing in the video is that if you do not refresh the page, the hotspot will return to the place where it originally was. and this does not suit me at all. in addition, it is advisable not to use the position change. Thank you in advance. I hope the video will be visible and will be played

Here is the part of the code that is responsible for moving the hotspot:
function moveHS(event) {
  contextMenu.style.display = "none";
  if (editMode && moveHotspot) {
    const hotspot = event.target;
    const currentSlot = event.target.getAttribute("slot");
    async function onMouseMove(moveEvent) {
      let newPos = modelViewer.positionAndNormalFromPoint(moveEvent.clientX, moveEvent.clientY);
      const updatedProperties = {
        name: currentSlot,
        position: `${newPos.position.x} ${newPos.position.y} ${newPos.position.z}`,
      };
      let newSur = modelViewer.surfaceFromPoint(moveEvent.clientX, moveEvent.clientY);
      console.log("-----------------------------");
      console.log(newSur);
      await modelViewer.updateComplete;
      hotspot.setAttribute('data-surface', newSur);
      const updatedSurface = {
        name: currentSlot,
        surface: `${newSur}`,
      };
      console.log(modelViewer.queryHotspot(currentSlot))
      modelViewer.updateHotspot(updatedProperties);
      modelViewer.updateHotspot(updatedSurface);

    }
    function onMouseUp() {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
      saveButtons();
    }

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }
}


https://github.com/user-attachments/assets/e761228b-82a3-400d-979c-b1ca77748830

**Reply by elalish** [reply]
It looks like you're quite close - I think the problem is you're trying to specify both the position/normal *and* the surface, but the API is designed to be one or the other. For your application you want a surface hotspot, not an absolute position hotspot.

**Reply by bigbea1r** [reply]
Thanks for the answer, but .positionAndNormalFromPoint was added for clarity, to see that Hotspot can move dynamically. If we leave the code as shown below, then the dot remains in place, but the data-surface value is constantly updated, but for some reason these updates are not assigned dynamically, but only after the page itself is updated.  I'll leave the video below. Thank you in advance

function moveHS(event) {
   
  if (editMode && onMoveHotspot) {
    const hotspot = event.target;
    const currentSlot = event.target.getAttribute("slot");
    async function onMouseMove(moveEvent) {
      let newSur = modelViewer.surfaceFromPoint(moveEvent.clientX, moveEvent.clientY);
      await modelViewer.updateComplete;
      hotspot.setAttribute("data-surface", newSur);
      const updatedSurface = {
        name: currentSlot,
        surface: `${newSur}`,
};
      modelViewer.updateHotspot(updatedSurface);
    }
    function onMouseUp() {
      document.removeEventListener("mousemove", onMouseMove);
      document.removeEventListener("mouseup", onMouseUp);
      saveButtons();
    }

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  }
}

https://github.com/user-attachments/assets/4a9945d3-f6a5-40d4-961e-8e59c9d77195

**Reply by elalish** [reply]
You shouldn't need `setAttribute` - `updateHotspot` should be sufficient. I would guess the bug is in your `editMode`/`onMoveHotspot` logic.

**Reply by bigbea1r** [reply]
I'm sorry that I bother you, but it still doesn't work, I simplified the code as much as possible, leaving only the transfer of the hotspot with the help of the left mouse button, but the value itself is updated, but the hotspot itself does not move around the model and does not update the value of the data-surface. And they would have helped out a lot more, what each coordinate in the data-surface means.

const modelViewer = document.querySelector("model-viewer");

function createHotspot() { 
    const hotspot = document.createElement("button"); 
    const hotspotDiv = document.createElement("div"); 
   
    hotspot.setAttribute("class", "Hotspot"); 
    hotspot.setAttribute("slot", `hotspot-1`); 
    hotspot.setAttribute("data-surface", '86 0 221 242 211 0.239 0.186 0.575'); 
    hotspot.setAttribute("data-visibility-attribute", "visible"); 
    hotspot.setAttribute("data-animation", 'Left'); 
    hotspotDiv.setAttribute("class", "HotspotAnnotation"); 
    hotspotDiv.textContent = 'Left door'; 
    hotspot.appendChild(hotspotDiv); 
 
    hotspot.addEventListener("mousedown", moveHS); 
   
    modelViewer.appendChild(hotspot);

    return hotspot;
}

const hotspot = createHotspot();
const observer = new MutationObserver((mutationsList) => { 
  for (let mutation of mutationsList) { 
    if (mutation.type === 'attributes' && mutation.attributeName === 'data-surface') { 
      console.log("data-surface изменен на:", mutation.target.getAttribute('data-surface')); 
    } 
  } 
});

observer.observe(hotspot, { attributes: true });

function moveHS(event) { 
  function onMouseMove(moveEvent) { 
    const newSurface = modelViewer.surfaceFromPoint(moveEvent.clientX, moveEvent.clientY); 
    if (newSurface) { 
      const updatedSurface = { 
        name: 'hotspot-1', 
        surface: `${newSurface}`, 
      }; 

      console.log("Новая поверхность:", newSurface); 

      modelViewer.updateHotspot(updatedSurface);

      console.log("data-surface изменен на:", newSurface); 
    } 
  } 
 
  document.addEventListener("mousemove", onMouseMove); 
 
  document.addEventListener("mouseup", () => { 
    document.removeEventListener("mousemove", onMouseMove); 
  }, { once: true }); 
}

https://github.com/user-attachments/assets/0dfe02f0-727c-4df5-bf2f-d620f6593995

**Reply by elalish** [reply]
There's really no way to know where the bug is without a link to the live site for us to test. As for what data-surface means - it's `[meshID, primitiveID, triangleIndex, barycentric a, barycentric b, barycentric c]`.

**Reply by bigbea1r** [reply]
> There's really no way to know where the bug is without a link to the live site for us to test. As for what data-surface means - it's `[meshID, primitiveID, triangleIndex, barycentric a, barycentric b, barycentric c]`.

Good time of day. If I understood correctly, then here is a link to a mini-project with this problem. If anything, I run it using the utility in VS code - Life Server
https://drive.google.com/drive/folders/1cBzPc7u0-NxfMdoKl8o-OhKuJv_wBVDR?usp=drive_link

**Reply by elalish** [reply]
Nope, I mean an actual deployed website - try copying that into Glitch or JSbin or something.

**Reply by bigbea1r** [reply]
> Нет, я имею в виду реально развернутый веб-сайт — попробуйте скопировать его в Glitch, JSbin или что-то в этом роде.

https://jsbin.com/gucikus/edit?html,css,js,console,output

**Reply by bigbea1r** [reply]
> > Нет, я имею в виду реально развернутый веб-сайт — попробуйте скопировать его в Glitch, JSbin или что-то в этом роде.
> 

https://vine-thirsty-quotation.glitch.me
https://glitch.com/edit/#!/vine-thirsty-quotation

**Reply by elalish** [reply]
Thank you for the clear reproduction example! I found the problem, and it is indeed a MV bug, and I think I can fix it with one line.

**Reply by elalish** [reply, marked as answer]
Okay, turned out to need a bit more than a one-liner, but our logic is quite a bit more robust now: #4867. I verified that it fixes your Glitch, and improved our automated test. Still has a funny CSS offset, but that shouldn't be hard for you to fix.

**Reply by bigbea1r** [reply]
> Okay, turned out to need a bit more than a one-liner, but our logic is quite a bit more robust now: #4867. I verified that it fixes your Glitch, and improved our automated test. Still has a funny CSS offset, but that shouldn't be hard for you to fix.

Thank you very much in advance. I'm going to try it now
